### ISSUE TITLE
Change data generator to mnist-rot

### ISSUE THREAD

**Issue by ymcasky** [jxgu1016/Gabor_CNN_PyTorch#6]
Hi,

I using data generator as ORN as following in main.py.

```
#train_dataset = datasets.MNIST(root=args.dataset_dir, train=True, 
#    				download=True, transform=train_transform)
#test_dataset = datasets.MNIST(root=args.dataset_dir, train=False, 
#                    download=True,transform=test_transform)
#
#train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=args.batch_size,
#                                num_workers=args.workers, pin_memory=True, shuffle=True)
#test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=args.batch_size,
#                                num_workers=args.workers, pin_memory=True, shuffle=True)


import math
import numbers
import random
from PIL import Image, ImageOps
# custom transform
class RandomRotate(object):
    """Rotate the given PIL.Image counter clockwise around its centre by a random degree
    (drawn uniformly) within angle_range. angle_range is a tuple (angle_min, angle_max).
    Empty region will be padded with color specified in fill."""
    def __init__(self, angle_range=(-180,180), fill='black'):
        assert isinstance(angle_range, tuple) and len(angle_range) == 2 and angle_range[0] <= angle_range[1]
        assert isinstance(fill, numbers.Number) or isinstance(fill, str) or isinstance(fill, tuple)
        self.angle_range = angle_range
        self.fill = fill

    def __call__(self, img):
        angle_min, angle_max = self.angle_range
        angle = angle_min + random.random() * (angle_max - angle_min)
        theta = math.radians(angle)
        w, h = img.size
        diameter = math.sqrt(w * w + h * h)
        theta_0 = math.atan(float(h) / w)
        w_new = diameter * max(abs(math.cos(theta-theta_0)), abs(math.cos(theta+theta_0)))
        h_new = diameter * max(abs(math.sin(theta-theta_0)), abs(math.sin(theta+theta_0)))
        pad = math.ceil(max(w_new - w, h_new - h) / 2)
        img = ImageOps.expand(img, border=int(pad), fill=self.fill)
        img = img.rotate(angle, resample=Image.BICUBIC)
        return img.crop((pad, pad, w + pad, h + pad))


train_loader = torch.utils.data.DataLoader(
    datasets.MNIST(root=args.dataset_dir, train=True, download=True,
                   transform=transforms.Compose([
                       transforms.Scale(32),
                       RandomRotate((-180, 180)),
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ])), batch_size=args.batch_size,
                                num_workers=args.workers, pin_memory=True, shuffle=True)
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST(root=args.dataset_dir, train=False, transform=transforms.Compose([
                       transforms.Scale(32),
                       RandomRotate((-180, 180)),
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ])), batch_size=args.batch_size,
                                num_workers=args.workers, pin_memory=True, shuffle=True)
```
The original data size is 28x28.
The original size after x = self.model(x) is 128x320x1x1.
After x = x.view(-1, 80*2, self.channel) is 128x80x4

The mnist-rot data is padding to 32x32.
The size after x = self.model(x) is 128x320x2x2.
After x = x.view(-1, 80*2, self.channel) is 512x80x4 which change the batch size!!

Can you tell me how to fix it or just simply change the padding number in net_factory?
Thank you!

**Comment by jxgu1016** [comment]
What about ```x = x.view(-1, 80*4, self.channel)```

**Comment by jxgu1016** [comment]
1. To change the padding number **2** to **0** in the first or second conv layer will help. 
2. Resize the input images to **28*28** after rotation using ```transforms.Scale(28)``` (recommended).

**Comment by ymcasky** [comment]
I using
1.  change first 2 layer's padding size to 1
```
GConv(1, 10, 5, padding=1, stride=1, M=channel, nScale=1, bias=False, expand=True),
            nn.BatchNorm2d(10*channel),
            nn.ReLU(inplace=True),

            GConv(10, 20, 5, padding=1, stride=1, M=channel, nScale=2, bias=False),
            nn.BatchNorm2d(20*channel),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(2,2),
```

2. change first 2 layer's padding size to 1 + all fc connecting

```
GConv(1, 10, 5, padding=1, stride=1, M=channel, nScale=1, bias=False, expand=True),
            nn.BatchNorm2d(10*channel),
            nn.ReLU(inplace=True),

            GConv(10, 20, 5, padding=1, stride=1, M=channel, nScale=2, bias=False),
            nn.BatchNorm2d(20*channel),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(2,2),

+ self.fc1 = nn.Linear(80*4, 1024)

+ x = x.view(-1, 80 * self.channel)
#        x = x.view(-1, 80, self.channel)
#        x = torch.max(x, 2)[0]

```
3. As your suggestion

```
GConv(1, 10, 5, padding=2, stride=1, M=channel, nScale=1, bias=False, expand=True),
            nn.BatchNorm2d(10*channel),
            nn.ReLU(inplace=True),

            GConv(10, 20, 5, padding=2, stride=1, M=channel, nScale=2, bias=False),
            nn.BatchNorm2d(20*channel),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(2,2),
 + self.fc1 = nn.Linear(80*4, 1024)

+ x = x.view(-1, 80*4, self.channel)
```


But the testing accuracy can't reach above 99% which ORN can :(

I will try your recommendation next!
